### PR TITLE
Optiguide

### PR DESCRIPTION
**Pull request by BeibinLi**
## Why are these changes needed?
Adding an application on OptiGuide, a framework to answer supply chain optimization questions through coding.

Create a OptiGuideAgent, a notebook, and an option for setup.py

## Checks

- [ ] I've included any doc changes needed for https://microsoft.github.io/FLAML/. See https://microsoft.github.io/FLAML/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [ ] I've made sure all auto checks have passed.

### CODE REVIEW

**Review by sonichi — COMMENTED** [review]
The notebook looks very nice. Please review #1148 before addressing the comments here because some comments are related to that.

**Comment by BeibinLi** [comment]
@sonichi   Thanks for these suggestions! I have addressed most concerns, except the following:

1. The "    except Exception as e:" code line is designed to catch all possible strange errors could occur during execution (e.g., variable name not exist, key error, etc.). So, let's keep it this way. 
2. Not sure how to replace "receive()" function with "generate_reply()". The "generate_reply" will not triggered in the notebook.
3. I agree "_replace" can be moved to code_utils.py. I can open another PR later to address this issue.
4.  termcolor is a lightweight package, which is designed to color output in prints and terminal. It is very popular.

**Comment by sonichi** [comment]
> 

1.The line I commented for the Exception is not doing execution. It's check the status after execution. So I think the exception type is more limited.
2.`generate_reply` is triggered by `receive` automatically. If you override `generate_reply` instead of `receive`, it should be called by ResponsiveAgent's `receive`.
3. Sounds good.
4. Thanks for the input.

File: flaml/autogen/agent/opti_guide.py [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+    and two assistant agents (pencil and shield).
+    """
+
+    def __init__(self, name, source_code, doc_str="", example_qa="", **config):
+        """
+        Args:
+            name (str): agent name.
+            source_code (str): The original source code to run.
+            system_message (str): system message to be sent to the agent.
+            **config (dict): other configurations allowed in
```
[the remark this answers is not in the document]
Comment (sonichi) [reply flaml/autogen/agent/opti_guide.py]: This comment hasn't been addressed. The docstr here is not accurate.

**Pull request closed without merging**